prime-router/src/main/kotlin/serializers/Hl7Serializer.kt

```diff
@@ -58,10 +78,181 @@ class Hl7Serializer(val metadata: Metadata) {
+                deconstructStringMessage()
+            }
+            nextMessage.append("$it\r")
+        }
+
+        // catch the last message
+        if (nextMessage.isNotEmpty()) {
+            deconstructStringMessage()
+        }
+
+        return Hl7Mapping(mappedRows, rowResults, errors, warnings)
+    }
+
+    fun convertMessageToMap(message: String, schema: Schema): Hl7Serializer.RowResult {
+        // safely merge into the set. might not be necessary
+        fun mergeIntoMappedRows(mappedRows: MutableMap<String, MutableSet<String>>, key: String, value: String) {
+            if (!mappedRows.containsKey(key)) error("Map doesn't contain key $key")
+            // get the existing values
+            val existingValues = mappedRows[key] ?: emptySet()
+            // if the existing value doesn't exist, add it in
+            if (!existingValues.contains(value)) {
+                mappedRows[key]?.add(value)
+            }
+        }
+        // query the terser and get a value
+        fun queryTerserForValue(
+            terser: Terser,
+            terserSpec: String,
+            elementName: String,
+            mappedRows: MutableMap<String, MutableSet<String>>,
+            errors: MutableList<String>,
+            warnings: MutableList<String>
+        ) {
+            val parsedValue = try {
+                terser.get(terserSpec)
+            } catch (e: HL7Exception) {
+                errors.add("Exception for $terserSpec: ${e.message}")
+                null
+            }
+            // add the rows
+            if (parsedValue.isNullOrEmpty()) {
+                warnings.add("Blank for $terserSpec")
+            }
+            mergeIntoMappedRows(mappedRows, elementName, parsedValue ?: "")
+        }
+        val errors = mutableListOf<String>()
+        val warnings = mutableListOf<String>()
+        // key of the map is the column header, list is the values in the column
+        val mappedRows: MutableMap<String, MutableSet<String>> = mutableMapOf()
+        val mcf = CanonicalModelClassFactory("2.5.1")
+        hapiContext.modelClassFactory = mcf
+        val parser = hapiContext.pipeParser
+        val reg = "(\r|\n)".toRegex()
+        val cleanedMessage = reg.replace(message, "\r")
+        val hapiMsg = parser.parse(cleanedMessage)
+        val terser = Terser(hapiMsg)
+        schema.elements.forEach {
+            if (!mappedRows.containsKey(it.name))
+                mappedRows[it.name] = mutableSetOf()
+
+            if (it.hl7Field.isNullOrEmpty() && it.hl7OutputFields.isNullOrEmpty()) {
+                mappedRows[it.name]?.add("")
+                return@forEach
+            }
+
+            if (!it.hl7Field.isNullOrEmpty()) {
+                val terserSpec = if (it.hl7Field.startsWith("MSH")) {
+                    "/${it.hl7Field}"
+                } else if (it.hl7Field == "AOE") {
+                    val question = it.hl7AOEQuestion!!
+                    val countObservations = 10
```

Comment:
I tried multiple ways to make this dynamic but every code sample using HAPI to get the number of OBX segments failed me. For now this will do.